#716 Utilise digest authentication in curl commands

ryanemerson opened the pull request:
It's no longer possible to utilise BASIC authentication in our curl commands as the image only allows DIGEST authentication by default. Prior to CR1 this was not respected by the server, however this has been fixed by [ISPN-12576](https://issues.redhat.com/browse/ISPN-12576), therefore it's necessary to utilise digest authentication in our curl commands.

dmvolod commented on pkg/infinispan/client/http/curl/curl.go:
```diff
@@ -51,7 +51,7 @@ func (c *CurlClient) Post(podName, path, payload string, headers map[string]stri
 	httpURL := fmt.Sprintf("%s://%s:%d/%s", c.config.Protocol, podName, consts.InfinispanPort, path)
 	user := fmt.Sprintf("-u %v:%v", c.credentials.Username, c.credentials.Password)
-	curl := fmt.Sprintf("curl -i --insecure --http1.1 %s %s %s %s", user, headerString(headers), strings.Join(args, " "), httpURL)
+	curl := fmt.Sprintf("curl -i --insecure --digest --http1.1 %s %s %s %s", user, headerString(headers), strings.Join(args, " "), httpURL)
```
@ryanemerson is this working for 11.x images? I'm worries about upgrade case

ryanemerson replied:
Yep. Digest was supported and enabled by default in 11.x. I have manually ran the curl commands against the docker image and it works as expected.

I haven't tried the pre CR1 images, however they should work for the same reason. Regardless CR1 images were Dev releases so we provide no guarantees of compatibility.

ryanemerson commented on pkg/infinispan/client/http/curl/curl.go:
```diff
@@ -64,7 +64,17 @@ func (c *CurlClient) executeCurlCommand(podName string, path string, headers map
+		return nil, err, ""
+	}
+
+	if rsp.StatusCode != http.StatusUnauthorized {
```
I think we should handle the no auth case on a separate path as the `curl` command will also be different. For example:

```go
func (c *CurlClient) executeCurlCommand(podName string, path string, headers map[string]string, args ...string) (*http.Response, error, string) {
	httpURL := fmt.Sprintf("%s://%s:%d/%s", c.config.Protocol, podName, consts.InfinispanPort, path)
        	httpURL := fmt.Sprintf("%s://%s:%d/%s", c.config.Protocol, podName, consts.InfinispanPort, path)
	if c.credentials != nil {
		c.executeDigest(httpURL, headers)
	} else {
		c.execute(httpURL, headers)
	}
}
```

ryanemerson commented:
I've added an additional commit which introduces a very basic http lib with digest support and updated our tests. The upshot of this is that it cleans up the code a bit so we don't have to pass user/pass and protocol strings between all of the methods.

rigazilla commented:
merged, thanks @ryanemerson !

ryanemerson commented:
Thanks for taking the time to diagnose why CR1 was failing @dmvolod, much appreciated.
